File: android/src/main/kotlin/ai/decart/vton/flutter/VtonVideoPlatformView.kt
package ai.decart.vton.flutter

import ai.decart.sdk.realtime.RealtimeMediaStream
import android.content.Context
import android.view.View
import android.widget.FrameLayout
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.livekit.android.renderer.TextureViewRenderer
import io.livekit.android.room.Room
import io.livekit.android.room.track.VideoTrack
import livekit.org.webrtc.RendererCommon

/** Which track a video view is bound to. */
internal enum class VtonVideoSource { REMOTE, LOCAL }

/**
 * Owns the set of live video views and keeps them bound to the current tracks.
 *
 * Rebinding happens **here**, natively, rather than by telling Dart to tear
 * down and rebuild the platform view. The Decart SDK emits a fresh
 * [RealtimeMediaStream] after every automatic reconnect; round-tripping that
 * through Dart would mean destroying and recreating a platform view (a visible
 * black flash, plus a window of dropped frames) for something the native side
 * can fix in place.
 */
internal class VtonVideoViewRegistry {

    private val views = mutableListOf<VtonVideoPlatformView>()
    private var remote: RealtimeMediaStream? = null
    private var local: RealtimeMediaStream? = null

    fun register(view: VtonVideoPlatformView) {
        views.add(view)
        view.bind(streamFor(view.source))
    }

    fun unregister(view: VtonVideoPlatformView) {
        views.remove(view)
    }

    fun setRemote(stream: RealtimeMediaStream?) {
        remote = stream
        views.filter { it.source == VtonVideoSource.REMOTE }.forEach { it.bind(stream) }
    }

    fun setLocal(stream: RealtimeMediaStream?) {
        local = stream
        views.filter { it.source == VtonVideoSource.LOCAL }.forEach { it.bind(stream) }
    }

    /** Unbinds everything. Called on disconnect so views do not hold dead tracks. */
    fun clearStreams() {
        remote = null
        local = null
        views.forEach { it.bind(null) }
    }

    private fun streamFor(source: VtonVideoSource): RealtimeMediaStream? =
        if (source == VtonVideoSource.REMOTE) remote else local
}

/** Creates [VtonVideoPlatformView]s for the Flutter platform-view system. */
internal class VtonVideoViewFactory(
    private val registry: VtonVideoViewRegistry,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val params = (args as? Map<String, Any?>) ?: emptyMap()
        return VtonVideoPlatformView(
            context = context,
            source = if (params["source"] == "local") {
                VtonVideoSource.LOCAL
            } else {
                VtonVideoSource.REMOTE
            },
            scaleToFill = params["fit"] != "contain",
            mirror = params["mirror"] as? Boolean ?: false,
            registry = registry,
        )
    }
}

/**
 * A LiveKit [TextureViewRenderer] hosted inside a [FrameLayout], exposed to
 * Flutter as a platform view.
 *
 * The [FrameLayout] wrapper exists so the renderer can be destroyed and
 * recreated (which is necessary when the owning [Room] — and therefore the
 * `EglBase` — changes) without the platform view itself going away.
 * Re-`init`-ing an existing renderer against a different `EglBase` is not
 * reliable; the LiveKit sample works around it by recreating the whole
 * composable, and this is the equivalent.
 */
internal class VtonVideoPlatformView(
    private val context: Context,
    val source: VtonVideoSource,
    private val scaleToFill: Boolean,
    private val mirror: Boolean,
    private val registry: VtonVideoViewRegistry,
) : PlatformView {

    private val container = FrameLayout(context).apply {
        layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT,
        )
    }

    private var renderer: TextureViewRenderer? = null
    private var boundRoom: Room? = null
    private var boundTrack: VideoTrack? = null
    private var disposed = false

    init {
        registry.register(this)
    }

    /**
     * Points this view at [stream], creating or recreating the underlying
     * renderer as needed. Passing `null` detaches and releases.
     *
     * Must be called on the main thread — [TextureViewRenderer.init] touches
     * the view hierarchy. Every call site is on `Dispatchers.Main` or inside
     * the main-looper post in [VtonEventDispatcher].
     */
    fun bind(stream: RealtimeMediaStream?) {
        if (disposed) return

        val room = stream?.room
        val track = stream?.videoTrack
        if (room == null || track == null) {
            detachTrack()
            return
        }

        if (boundRoom !== room) {
            releaseRenderer()
            createRenderer(room)
            boundRoom = room
        }

        val activeRenderer = renderer ?: return
        if (boundTrack !== track) {
            boundTrack?.let { runCatching { it.removeRenderer(activeRenderer) } }
            runCatching { track.addRenderer(activeRenderer) }
            boundTrack = track
        }
    }

    private fun createRenderer(room: Room) {
        val created = TextureViewRenderer(context)
        created.init(room.lkObjects.eglBase.eglBaseContext, null)
        created.setEnableHardwareScaler(true)
        created.setMirror(mirror)
        created.setScalingType(
            if (scaleToFill) {
                RendererCommon.ScalingType.SCALE_ASPECT_FILL
            } else {
                RendererCommon.ScalingType.SCALE_ASPECT_FIT
            },
        )
        created.layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT,
        )
        container.addView(created)
        renderer = created
    }

    private fun detachTrack() {
        val activeRenderer = renderer ?: return
        boundTrack?.let { runCatching { it.removeRenderer(activeRenderer) } }
        boundTrack = null
    }

    private fun releaseRenderer() {
        detachTrack()
        renderer?.let { active ->
            container.removeView(active)
            runCatching { active.release() }
        }
        renderer = null
        boundRoom = null
    }

    override fun getView(): View = container

    override fun dispose() {
        if (disposed) return
        disposed = true
        registry.unregister(this)
        releaseRenderer()
        container.removeAllViews()
    }
}

File: ios/decart_vton_flutter/Sources/decart_vton_flutter/VtonVideoPlatformView.swift
import DecartSDK
import Flutter
@preconcurrency import LiveKit
import UIKit

/// Which track a video view is bound to.
enum VtonVideoSource: Sendable {
    case remote
    case local
}

/// Owns the live video views and keeps them bound to the current tracks.
///
/// Rebinding happens here, natively, rather than by asking Dart to rebuild the
/// platform view. The SDK emits a fresh `RealtimeMediaStream` after every
/// automatic reconnect; recreating a `UiKitView` for that would mean a visible
/// black flash for something that is a one-line property assignment on
/// LiveKit's `VideoView`.
///
/// Mirrors `VtonVideoViewRegistry` in the Kotlin sources — keep the two in step.
///
/// ## Concurrency
///
/// Marked `@unchecked Sendable` rather than `@MainActor`. Everything that
/// touches this type already runs on the platform thread: platform views are
/// created there by the Flutter engine, and every `setRemote`/`setLocal` call
/// originates from `VtonSessionController`, which *is* `@MainActor`. Making the
/// registry itself main-actor-isolated would force `MainActor.assumeIsolated`
/// hops in `VtonVideoViewFactory.create` (a nonisolated ObjC protocol witness)
/// and in `FlutterPlatformView.view()`, neither of which can return a
/// non-`Sendable` `UIView` through `assumeIsolated`.
final class VtonVideoViewRegistry: @unchecked Sendable {

    /// Weak so a view that Flutter has torn down drops out on its own.
    ///
    /// `FlutterPlatformView` has no `dispose` hook (unlike Android's
    /// `PlatformView.dispose`), and unregistering from `deinit` would mean
    /// escaping `self` out of a deinitialiser — undefined behaviour. A weak
    /// table sidesteps the whole problem.
    private let views = NSHashTable<VtonVideoPlatformView>.weakObjects()

    private var remote: RealtimeMediaStream?
    private var local: RealtimeMediaStream?

    func register(_ view: VtonVideoPlatformView) {
        views.add(view)
        view.bind(stream(for: view.source))
    }

    func setRemote(_ stream: RealtimeMediaStream?) {
        remote = stream
        for view in views.allObjects where view.source == .remote {
            view.bind(stream)
        }
    }

    func setLocal(_ stream: RealtimeMediaStream?) {
        local = stream
        for view in views.allObjects where view.source == .local {
            view.bind(stream)
        }
    }

    /// Unbinds everything, so views do not hold dead tracks after a disconnect.
    func clearStreams() {
        remote = nil
        local = nil
        for view in views.allObjects {
            view.bind(nil)
        }
    }

    private func stream(for source: VtonVideoSource) -> RealtimeMediaStream? {
        source == .remote ? remote : local
    }
}

/// Creates `VtonVideoPlatformView`s for the Flutter platform-view system.
final class VtonVideoViewFactory: NSObject, FlutterPlatformViewFactory {

    private let registry: VtonVideoViewRegistry

    init(registry: VtonVideoViewRegistry) {
        self.registry = registry
        super.init()
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        FlutterStandardMessageCodec.sharedInstance()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        let params = args as? [String: Any] ?? [:]
        return VtonVideoPlatformView(
            frame: frame,
            source: (params["source"] as? String) == "local" ? .local : .remote,
            scaleToFill: (params["fit"] as? String) != "contain",
            mirror: (params["mirror"] as? Bool) ?? false,
            registry: registry
        )
    }
}

/// A LiveKit `VideoView` exposed to Flutter as a `UiKitView`.
///
/// `VideoView` handles renderer lifecycle internally (unlike the Android side,
/// where the renderer must be recreated when the owning `Room`'s `EglBase`
/// changes), so binding is just an assignment.
///
/// `@unchecked Sendable` for the same reason as the registry: this type is only
/// ever constructed and mutated on the platform thread, and `view()` is a
/// nonisolated ObjC protocol requirement returning a non-`Sendable` `UIView`,
/// which no actor-isolation annotation can express cleanly.
final class VtonVideoPlatformView: NSObject, FlutterPlatformView, @unchecked Sendable {

    let source: VtonVideoSource

    private let container: UIView
    private let videoView: VideoView

    init(
        frame: CGRect,
        source: VtonVideoSource,
        scaleToFill: Bool,
        mirror: Bool,
        registry: VtonVideoViewRegistry
    ) {
        self.source = source

        container = UIView(frame: frame)
        container.backgroundColor = .clear
        container.clipsToBounds = true

        videoView = VideoView(frame: container.bounds)
        videoView.layoutMode = scaleToFill ? .fill : .fit
        videoView.mirrorMode = mirror ? .mirror : .off
        videoView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        container.addSubview(videoView)

        super.init()
        registry.register(self)
    }

    func view() -> UIView { container }

    /// Points this view at `stream`, or clears it when `stream` is nil.
    ///
    /// `VideoView` releases its renderer when `track` is set to nil, so this is
    /// also the teardown path — the registry calls it with nil on disconnect.
    func bind(_ stream: RealtimeMediaStream?) {
        videoView.track = stream?.videoTrack
    }
}
